oak-jcr/src/main/java/org/apache/jackrabbit/oak/jcr/session/SessionImpl.java

```diff
@@ -773,6 +785,122 @@ public UserManager getUserManager() throws RepositoryException {
+                        @Override
+                        public String getUploadToken() {
+                            return upload.getUploadToken();
+                        }
+
+                        @Override
+                        public long getMinPartSize() {
+                            return upload.getMinPartSize();
+                        }
+
+                        @Override
+                        public long getMaxPartSize() {
+                            return upload.getMaxPartSize();
+                        }
+
+                        @Nonnull
+                        @Override
+                        public Collection<URL> getUploadURLs() {
+                            return upload.getUploadURLs();
+                        }
+                    };
+                }
+                return null;
+            }
+        });
+    }
+
+    @Nonnull
+    @Override
+    public Binary completeHttpUpload(String uploadToken) throws RepositoryException {
+        return sd.perform(new ReadOperation<Binary>("completeHttpUpload") {
+
+            @Nonnull
+            @Override
+            public Binary perform() throws RepositoryException {
+
+                HttpBlobProvider httpBlobProvider = getHttpBlobProvider();
+                if (httpBlobProvider == null) {
+                    throw new RepositoryException("HTTP binary upload not supported");
+                }
+                Blob blob = httpBlobProvider.completeHttpUpload(uploadToken);
+                if (blob == null) {
+                    throw new RepositoryException("HTTP binary upload could not be completed");
+                }
+
+                ValueFactoryImpl vf = (ValueFactoryImpl) sessionContext.getValueFactory();
+                return vf.createBinary(blob);
+            }
+        });
+    }
+
+    @Nullable
+    @Override
+    public URL getHttpDownloadURL(Binary binary) {
+        return sd.safePerformNullable(new ReadOperation<URL>("getHttpDownloadURL") {
+
+            @Nullable
+            @Override
+            public URL performNullable() throws RepositoryException {
+                HttpBlobProvider httpBlobProvider = getHttpBlobProvider();
+                if (httpBlobProvider == null) {
+                    return null;
+                }
+                if (binary instanceof ReferenceBinary) {
+                    if (null == ((ReferenceBinary) binary).getReference()) {
+                        // Binary is inlined, we cannot return a URL for it
+                        return null;
+                    }
+
+                    // ValueFactoryImpl.getBlobId() will only return a blobId for a BinaryImpl, which
+                    // a client cannot spoof, so we know that the id in question is valid and can be
+                    // trusted, so we can safely give out a URL to the binary for downloading.
+                    String blobId = ((ValueFactoryImpl) sessionContext.getValueFactory()).getBlobId(binary);
```

Comment:
Regarding the `ValueFactoryImpl` cast, see my comment above in `completeHttpUpload()`.